lib/shelley/src/Cardano/Wallet/Shelley.hs

```diff
@@ -0,0 +1,454 @@
+import Cardano.Wallet.Api.Types
+    ( DecodeAddress, EncodeAddress )
+import Cardano.Wallet.DB.Sqlite
+    ( DefaultFieldValues (..), PersistState )
+import Cardano.Wallet.Logging
+    ( filterTraceSeverity, trMessageText )
+import Cardano.Wallet.Network
+    ( NetworkLayer (..) )
+import Cardano.Wallet.Primitive.AddressDerivation
+    ( Depth (..)
+    , NetworkDiscriminant (..)
+    , NetworkDiscriminantVal
+    , PaymentAddress
+    , PersistPrivateKey
+    , WalletKey
+    , networkDiscriminantVal
+    )
+import Cardano.Wallet.Primitive.AddressDerivation.Byron
+    ( ByronKey )
+import Cardano.Wallet.Primitive.AddressDerivation.Icarus
+    ( IcarusKey )
+import Cardano.Wallet.Primitive.AddressDerivation.Shelley
+    ( ShelleyKey )
+import Cardano.Wallet.Primitive.AddressDiscovery
+    ( IsOurs )
+import Cardano.Wallet.Primitive.AddressDiscovery.Random
+    ( RndState )
+import Cardano.Wallet.Primitive.AddressDiscovery.Sequential
+    ( SeqState )
+import Cardano.Wallet.Primitive.Types
+    ( Address
+    , Block
+    , BlockchainParameters (..)
+    , ChimericAccount
+    , GenesisBlockParameters (..)
+    , SyncTolerance
+    , WalletId
+    )
+import Cardano.Wallet.Registry
+    ( WorkerLog (..) )
+import Cardano.Wallet.Shelley.Api.Server
+    ( server )
+import Cardano.Wallet.Shelley.Compatibility
+    ( Shelley, ShelleyBlock, fromNetworkMagic, fromShelleyBlock )
+import Cardano.Wallet.Shelley.Network
+    ( NetworkLayerLog, withNetworkLayer )
+import Cardano.Wallet.Shelley.Transaction
+    ( newTransactionLayer )
+import Cardano.Wallet.Transaction
+    ( TransactionLayer )
+import Control.Applicative
+    ( Const (..) )
+import Control.Tracer
+    ( Tracer (..), nullTracer, traceWith )
+import Data.Function
+    ( (&) )
+import Data.Proxy
+    ( Proxy (..) )
+import Data.Text
+    ( Text )
+import Data.Text.Class
+    ( ToText (..) )
+import GHC.Generics
+    ( Generic )
+import Network.Ntp
+    ( NtpClient (..), NtpTrace (..), withWalletNtpClient )
+import Network.Socket
+    ( SockAddr, Socket, getSocketName )
+import Network.Wai.Handler.Warp
+    ( setBeforeMainLoop )
+import Network.Wai.Middleware.Logging
+    ( ApiLog )
+import Ouroboros.Network.CodecCBORTerm
+    ( CodecCBORTerm )
+import Ouroboros.Network.NodeToClient
+    ( NodeToClientVersionData (..) )
+import System.Exit
+    ( ExitCode (..) )
+import System.IOManager
+    ( withIOManager )
+
+import qualified Cardano.Wallet.Api.Server as Server
+import qualified Cardano.Wallet.DB.Sqlite as Sqlite
+import qualified Data.Text as T
+import qualified Network.Wai.Handler.Warp as Warp
+
+-- | Encapsulate a network discriminant and the necessary constraints it should
+-- satisfy.
+data SomeNetworkDiscriminant where
+    SomeNetworkDiscriminant
+        :: forall (n :: NetworkDiscriminant).
+            ( NetworkDiscriminantVal n
+            , PaymentAddress n IcarusKey
+            , PaymentAddress n ByronKey
+            , PaymentAddress n ShelleyKey
```

Comment:
This is incorrect at the moment, but we can solve later. We need to:

a) Rename 'ShelleyKey' to 'JormungandrKey'
b) Port https://github.com/input-output-hk/cardano-addresses/blob/master/src/Cardano/Address/Style/Shelley.hs